---
title: "Estimating Tree Similarity Using Permutational Hypothesis Testing"
author: "Jackson Turner"
date: "5/18/2022"
output: github_document
---

## Comparing Two Phylogenies Using Permutational Hypothesis Testing:

There are several ways to compare the phylogenies of two different organisms, and one way to do that is to use a permutational hypothesis test to evlaluate if two phylogenies share a significantly high or low number of clades. For the sake of this exercise, I'll be using host and endosymbiont phylogenies constructed from unpublished data using Dixidae (Diptera:Culicidae) CO1 genes and Rickettsial SNP data respectively. My H1 and H0 are as follows:

H1: Host and endosymbiont trees share a different number of clades than the host tree and randomly-generated trees.

H0: Host and endosymbiont trees share the same number of clades as the host tree and randomly-generated trees.

I'll test my hypothesis by simulating several random trees with the same tip labels as the host tree and extracting the number of shared clades -- as I continue to sample, the number of shared clades between these random trees and my actual tree will represent my hypothesis that host and endosymbiont trees are different. Then, I'll plot the number of shared clades as a histogram and then the number of shared clades between the actual host and endosymbiont trees as an abline to see where it falls. If it's within the range of a normal distribution then my hypothesis will supported! Though, I'll need to account for the fact that that the tree I have probably isn't the actual tree -- and the same is true for the endosymbiont tree.

Load our packages.

```{r setup, include=FALSE, warning=FALSE}
library(ape)
library(phytools)
library(geiger)
library(phangorn)
library(TreeSim)
library(tidyr)
library(phylometrics)
```

Load in the host (dixid) tree constructed from CO1 data using RAxML

```{r chunk_1}
dixid_tree<-read.tree("all_CO1_GTRCAT.TREE")
plot(dixid_tree)
```

Load in our endosymbiont trees. These are both generated from Rickettsial SNP data using vcf2poptree -- one is generated using UPGMA and the other using neighor-joining methods, both from the same SNP data.

```{r chunk_2}
UPGMA_endosymbiont_tree<-read.tree("Rickettsia_UPGMA.TREE")
plot(UPGMA_endosymbiont_tree)
NJ_endosymbiont_tree<-read.tree("Rickettsia_NJ.TREE")
plot(NJ_endosymbiont_tree)
```

Extract the names of the tips to build the random trees.

```{r chunk_3}
dixid_tip_label_names<-NJ_endosymbiont_tree$tip.label
```

As of 5/18/22 I don't have a host phylogeny from the same taxa as the endosymbiont trees, so I'm creating a new host tree using `rtree()` and the endosymbiont tip labels. 

```{r chunk_4}
set.seed(31)
dixid_tree<-rtree(n=length(dixid_tip_label_names),rooted=TRUE,tip.label=dixid_tip_label_names)
```

Load in helpful functions used to retrieve the shared clades between two trees (thanks Dr. Brian O'Meara!). 

```{r chunk_5}
GetTerminalDescendants <- function(tree) {
  return(Descendants(tree, ((1+ape::Ntip(tree)):(ape::Nnode(tree)+ape::Ntip(tree))), type="tips"))
}

ConvertTipVectorToString <- function(x) {
  return(paste0(sort(x), collapse="_"))
}
```

Retrieve the clades in the dixid tree in a usable format.

```{r chunk_6}
tree_clades <- simplify2array(lapply(GetTerminalDescendants(dixid_tree), ConvertTipVectorToString))
```

Create an empty vector for the number of overlapping trees, and then generate 10,000 random trees using `rcoal()`and retrieve the number of shared clades with the host tree

```{r chunk_7}
number_overlapping_clades <- as.vector(NULL)
for (i in 1:10000){
  temp_tree<-rcoal(n=length(dixid_tip_label_names),rooted=TRUE,tip.label=dixid_tip_label_names)
  temp_tree_clades <- simplify2array(lapply(GetTerminalDescendants(temp_tree), ConvertTipVectorToString))
  number_overlapping_clades[i] <- sum(temp_tree_clades %in% tree_clades)
}
```

How many clades, on average, do these random trees share with the host tree? This should strongly resemble a normal distribution because of the Central Limits Theorem.

**How to interpret these plots:**

**The thick dotted line represents two standard deviations away from the mean of the distribution, the thin line represents three standard deviations away from the mean of the distribution, and the blue line represents where the tested value falls**

I use a histogram for this plot but I use density plots for the rest since they would display similar results for this exercise.

```{r chunk_8, out.height="200%", out.width="200%"}
hist(number_overlapping_clades,xlab="Number of shared clades",main="")
abline(v=mean(number_overlapping_clades),lwd=3,col="blue")
clade_sd<-sd(number_overlapping_clades)
abline(v=(mean(number_overlapping_clades))-2*clade_sd,lwd=2,lty=2)
abline(v=(mean(number_overlapping_clades))+2*clade_sd,lwd=2,lty=2)
abline(v=(mean(number_overlapping_clades))-3*clade_sd,lwd=1,lty=2)
abline(v=(mean(number_overlapping_clades))+3*clade_sd,lwd=1,lty=2)
```

How many clades do our endosymbiont trees share with the host tree?

```{r chunk_9}
UPGMA_endosymbiont_tree_clades <- simplify2array(lapply(GetTerminalDescendants(UPGMA_endosymbiont_tree), ConvertTipVectorToString))
NJ_endosymbiont_tree_clades <- simplify2array(lapply(GetTerminalDescendants(NJ_endosymbiont_tree), ConvertTipVectorToString))

UPGMA_endosymbiont_tree_overlapping_clades<- sum(UPGMA_endosymbiont_tree_clades %in% tree_clades)
NJ_endosymbiont_tree_overlapping_clades<- sum(NJ_endosymbiont_tree_clades %in% tree_clades)
```

To evaluate whether the number of shared clades between the host and endosymbiont trees is significant, we'll use the number of overlapping clades we generated from our random trees earlier as a null hypothesis. Therefore, if the number of shared clades is much higher or lower than our null hypothesis, we can reject it, inferring that our endosymbiont tree is not one drawn from this distribution of randomly generated trees. Otherwise, we couldn't prove that the number of clades shared between these two trees is any different than if the endosymbiont tree was randomly generated.

```{r chunk_10, out.height="200%", out.width="200%"}
plot(density(number_overlapping_clades),xlab="Number of shared clades",main="")
abline(v=(mean(number_overlapping_clades))-2*clade_sd,lwd=2,lty=2)
abline(v=(mean(number_overlapping_clades))+2*clade_sd,lwd=2,lty=2)
abline(v=(mean(number_overlapping_clades))-3*clade_sd,lwd=1,lty=2)
abline(v=(mean(number_overlapping_clades))+3*clade_sd,lwd=1,lty=2)

abline(v=UPGMA_endosymbiont_tree_overlapping_clades,lwd=3,col="blue")
```

The UPGMA tree shares 9 clades with the host tree. Because this the number of clades shared between the UPGMA endosymbiont tree and the host tree falls within two standard deviations of the mean of randomly generated trees, we cannot reject our null hypothesis of this tree being a randomly-generated tree.

```{r chunk_11, out.height="200%", out.width="200%"}
plot(density(number_overlapping_clades),xlab="Number of shared clades",main="")
abline(v=(mean(number_overlapping_clades))-2*clade_sd,lwd=2,lty=2)
abline(v=(mean(number_overlapping_clades))+2*clade_sd,lwd=2,lty=2)
abline(v=(mean(number_overlapping_clades))-3*clade_sd,lwd=1,lty=2)
abline(v=(mean(number_overlapping_clades))+3*clade_sd,lwd=1,lty=2)

abline(v=NJ_endosymbiont_tree_overlapping_clades,lwd=3,col="blue")
```

The neighbor-joining endosymbiont tree also shares 11 clades with the host tree -- we similarly cannot reject our null hypothesis.

Now we'll test perform this test with other types of randomly generated trees (`rtree()` and `treesim()`) to see if we get similar results.

Using treesim to generate the random trees. The parameters I used for building the `treesim()` trees are those used in the help function. I also chose a random amount of starting individuals in each (invented for the sake of building trees using this function) state. 

```{r chunk_12, out.height="200%", out.width="200%"}
number_overlapping_clades <- as.vector(NULL)
for (i in 1:10000){
  random_1<-(runif(1,0,1)*72)+1
  random_1<-round(random_1,0)
  random_2<-74-random_1
  temp_tree<-treesim(pars=c(0.1,0.1,0.05,0.05,0.1,0.1),N0=random_1,N1=random_2,sampling.f=c(1,1))
  temp_tree$tip.label<-dixid_tip_label_names
  temp_tree_clades <- simplify2array(lapply(GetTerminalDescendants(temp_tree), ConvertTipVectorToString))
  number_overlapping_clades[i] <- sum(temp_tree_clades %in% tree_clades)
}

plot(density(number_overlapping_clades),xlim=c(0,20),xlab="Number of shared clades",main="")
clade_sd<-sd(number_overlapping_clades)
abline(v=(mean(number_overlapping_clades))-2*clade_sd,lwd=2,lty=2)
abline(v=(mean(number_overlapping_clades))+2*clade_sd,lwd=2,lty=2)
abline(v=(mean(number_overlapping_clades))-3*clade_sd,lwd=1,lty=2)
abline(v=(mean(number_overlapping_clades))+3*clade_sd,lwd=1,lty=2)

abline(v=UPGMA_endosymbiont_tree_overlapping_clades,lwd=3,col="blue")
```

The number of shared clades between the UPGMA endosymbiont tree and the host tree fall outside three standard deviations of the mean of the number shared clades between the randomly-generated trees using the treesim() function and the host tree. Therefore, we can reject our null hypothesis that the number of shared clades between our host and UPGMA endosymbiont trees is drawn from the distribution of shared clades between our host and treesim-generated trees.

Because we're using a tree generated using rtree() as our host tree, we would expect our host tree to be different.

```{r chunk_13, out.height="200%", out.width="200%"}
plot(density(number_overlapping_clades),xlim=c(0,12),xlab="Number of shared clades",main="")
abline(v=(mean(number_overlapping_clades))-2*clade_sd,lwd=2,lty=2)
abline(v=(mean(number_overlapping_clades))+2*clade_sd,lwd=2,lty=2)
abline(v=(mean(number_overlapping_clades))-3*clade_sd,lwd=1,lty=2)
abline(v=(mean(number_overlapping_clades))+3*clade_sd,lwd=1,lty=2)

abline(v=NJ_endosymbiont_tree_overlapping_clades,lwd=3,col="blue")
```

Because the number of shared clades between the host and NJ endosymbiont trees is greater than 3 standard deviations from the mean of shared clades between the host tree and treesim-generated trees, we can simililary reject our null hypothesis that the number of shared clades between our host and NJ endosymbiont trees is drawn from the distribution of shared clades between our host and treesim-generated trees.

Using rtree to generate random trees:

```{r chunk_14, out.height="200%", out.width="200%"}
number_overlapping_clades <- as.vector(NULL)
for (i in 1:10000){
  temp_tree<-rtree(n=length(dixid_tip_label_names),rooted=TRUE,tip.label=dixid_tip_label_names)
  temp_tree_clades <- simplify2array(lapply(GetTerminalDescendants(temp_tree), ConvertTipVectorToString))
  number_overlapping_clades[i] <- sum(temp_tree_clades %in% tree_clades)
}

plot(density(number_overlapping_clades),xlab="Number of shared clades",main="")
clade_sd<-sd(number_overlapping_clades)
abline(v=(mean(number_overlapping_clades))-2*clade_sd,lwd=2,lty=2)
abline(v=(mean(number_overlapping_clades))+2*clade_sd,lwd=2,lty=2)
abline(v=(mean(number_overlapping_clades))-3*clade_sd,lwd=1,lty=2)
abline(v=(mean(number_overlapping_clades))+3*clade_sd,lwd=1,lty=2)

abline(v=UPGMA_endosymbiont_tree_overlapping_clades,lwd=3,col="blue")

plot(density(number_overlapping_clades),xlab="Number of shared clades",main="")
abline(v=(mean(number_overlapping_clades))-2*clade_sd,lwd=2,lty=2)
abline(v=(mean(number_overlapping_clades))+2*clade_sd,lwd=2,lty=2)
abline(v=(mean(number_overlapping_clades))-3*clade_sd,lwd=1,lty=2)
abline(v=(mean(number_overlapping_clades))+3*clade_sd,lwd=1,lty=2)

abline(v=NJ_endosymbiont_tree_overlapping_clades,lwd=3,col="blue")
```

Our results of this test (and even the distribution generated) using `rtree()` are the same as using `rcoal()`.

Since the UPGMA and NJ trees share the exact same number of clades (11) with our host tree, are the clades they share the same ones? 

For the UPGMA endosymbiont tree:

```{r chunk_15}
UPGMA_endosymbiont_tree_shared_clades<- which(UPGMA_endosymbiont_tree_clades %in% tree_clades)
NJ_endosymbiont_tree_shared_clades<- which(NJ_endosymbiont_tree_clades %in% tree_clades)

UPGMA_shared_clades_pair<-(UPGMA_endosymbiont_tree_clades[UPGMA_endosymbiont_tree_shared_clades])
node_label_vec_UPGMA<-as.vector(NULL)
for (i in 1:length(UPGMA_shared_clades_pair)){
  
  temp_clade_vec<-as.numeric(unlist(strsplit(UPGMA_shared_clades_pair[i],"_")))
  temp_node_labels<-as.vector(NULL)
  
  for (j in 1:length(temp_clade_vec)){
    
    temp_node_labels[j]<-attr(((Descendants(UPGMA_endosymbiont_tree))),"labels")[temp_clade_vec[j]]
    temp_node_labels<-paste(temp_node_labels,collapse="_")
    
  }
  cat(temp_node_labels)
  node_label_vec_UPGMA<-append(node_label_vec_UPGMA,temp_node_labels,after=1)
}
node_label_vec_UPGMA
```

Now for the neighbor-joining endosymbiont tree.

```{r chunk_17}
NJ_shared_clades_pair<-(NJ_endosymbiont_tree_clades[UPGMA_endosymbiont_tree_shared_clades])
node_label_vec_NJ<-as.vector(NULL)
for (i in 1:length(NJ_shared_clades_pair)){
  
  temp_clade_vec<-as.numeric(unlist(strsplit(NJ_shared_clades_pair[i],"_")))
  temp_node_labels<-as.vector(NULL)
  
  for (j in 1:length(temp_clade_vec)){
    
    temp_node_labels[j]<-attr(((Descendants(NJ_endosymbiont_tree))),"labels")[temp_clade_vec[j]]
    temp_node_labels<-paste(temp_node_labels,collapse="_")
    
  }
  cat(temp_node_labels)
  node_label_vec_NJ<-append(node_label_vec_NJ,temp_node_labels,after=1)
}

intersect(node_label_vec_UPGMA,node_label_vec_NJ)
```

All three trees (host, UPGMA_symbiont, and NJ_symbiont) seem to only share two clades -- one clade between the Dlongistyla and Dmaculata samples and another between the DindianaS and Dixa_Suction_Cup samples. Though, it's worth noting that the NJ_symbiont tree shares much larger clades (as in, with many tips) with the host tree than the UPGMA_symbiont tree.

How many tips are in clades shared between the endosymbiont and host trees?

For the UPGMA tree:

```{r chunk_18}
node_label_vec_UPGMA_numbers<-as.vector(NULL)
for (i in 2:length(UPGMA_shared_clades_pair)){
  
  temp_clade_vec<-as.numeric(unlist(strsplit(UPGMA_shared_clades_pair[i],"_")))
  node_label_vec_UPGMA_numbers<-append(node_label_vec_UPGMA_numbers,temp_clade_vec,after=1)
}
unique(node_label_vec_UPGMA_numbers)
```

For the NJ tree:

```{r chunk_19}
node_label_vec_NJ_numbers<-as.vector(NULL)
for (i in 2:length(NJ_shared_clades_pair)){
  
  temp_clade_vec<-as.numeric(unlist(strsplit(NJ_shared_clades_pair[i],"_")))
  node_label_vec_NJ_numbers<-append(node_label_vec_NJ_numbers,temp_clade_vec,after=1)
}
unique(node_label_vec_NJ_numbers)
```

Vizualize the tips that are in shared clades using the `nodelabels()` function in the respective endosymbiont phylogenies. 

**Tips from a clade shared with the host tree are denoted by blue dots.**

For the UPGMA tree:

```{r chunk_20, out.height="200%", out.width="200%"}
plot(UPGMA_endosymbiont_tree)
nodelabels(node=node_label_vec_UPGMA_numbers,pch=19,col="blue1")
```

For the NJ tree:

```{r chunk_21, out.width="200%"}
plot(NJ_endosymbiont_tree)
nodelabels(node=node_label_vec_NJ_numbers,pch=19,col="blue1")
```

According to our plots, the NJ tree has a higher number of tips which share a clade with the host tree. While the structure of these clades is fundamentally different between the host and endosymbiont trees, we can infer that the NJ endosymbiont tree shares larger clades with the host tree than the UPGMA tree.

Is the number of tips of shared clades between the host and endosymbiont trees the same as randomly generated trees? We'll use another permutaitonal hypothesis test (H0: they aren't any different) to answer this question.

Use a for loop to generate all the random trees and gather the number of shared clades shared with the dixid tree. I'm choosing only to test using `rcoal()` to make this assignment legible. 

```{r chunk_22}
temp_tree_length_no<-as.vector(NULL)
for (i in 1:10000){
  temp_tree<-rcoal(n=length(dixid_tip_label_names),rooted=TRUE,tip.label=dixid_tip_label_names)
  temp_tree_clades <- simplify2array(lapply(GetTerminalDescendants(temp_tree), ConvertTipVectorToString))
  temp_tree_shared_clades<- which(temp_tree_clades %in% tree_clades)
  temp_shared_clades_pair<-(temp_tree_clades[temp_tree_shared_clades])
  node_label_vec_temp<-as.vector(NULL)
  for (j in 2:length(temp_shared_clades_pair)){
    
    temp_clade_vec<-as.numeric(unlist(strsplit(temp_shared_clades_pair[j],"_")))
    temp_node_labels<-as.vector(NULL)
    temp_node_labels_temp<-as.vector(NULL)
    
    for (k in 1:length(temp_clade_vec)){
      
      temp_node_labels_temp<-attr(((Descendants(temp_tree))),"labels")[temp_clade_vec[1]]
      temp_node_labels_new<-as.vector(NULL)
      for (m in 1:length(temp_node_labels_temp)){
        if (temp_node_labels_temp[m]!="NA"){
          temp_node_labels_new<-temp_node_labels_temp[m]
        }
      }
      temp_node_labels<-paste(temp_node_labels_new,collapse="_")
      
    }
    node_label_vec_temp<-append(node_label_vec_temp,temp_node_labels,after=1)
  }
  node_label_vec_temp_numbers<-as.vector(NULL)
  for (h in 1:length(temp_shared_clades_pair)){
    
    temp_clade_vec<-as.numeric(unlist(strsplit(temp_shared_clades_pair[i],"_")))
    node_label_vec_temp_numbers<-append(node_label_vec_temp_numbers,temp_clade_vec,after=1)
  }
  temp_tree_length_no[i]<-length(node_label_vec_temp_numbers)
}
```

Deleting the first two values because this loop appears to have a short burn-in for the first two steps where all values produced are outliers.

```{r chunk_23, out.height="200%", out.width="200%"}
temp_tree_length_no<-temp_tree_length_no[-c(1:2)]

plot(density(temp_tree_length_no),xlab="Number of tips within a shared clade",main="")
clade_sd<-sd(temp_tree_length_no)
abline(v=(mean(temp_tree_length_no))-2*clade_sd,lwd=2,lty=2)
abline(v=(mean(temp_tree_length_no))+2*clade_sd,lwd=2,lty=2)
abline(v=(mean(temp_tree_length_no))-3*clade_sd,lwd=1,lty=2)
abline(v=(mean(temp_tree_length_no))+3*clade_sd,lwd=1,lty=2)

abline(v=length(unique(node_label_vec_UPGMA_numbers)),lwd=3,col="blue")
```

According to this test, we cannot reject the null hypothesis. The number of tips shared between the UPGMA and host tree is not significantly different than the number of tips shared between randomly-generated trees and the host trees. 

```{r chunk_24, out.height="200%", out.width="200%"}
plot(density(temp_tree_length_no),xlab="Number of tips within a shared clade",main="")
clade_sd<-sd(temp_tree_length_no)
abline(v=(mean(temp_tree_length_no))-2*clade_sd,lwd=2,lty=2)
abline(v=(mean(temp_tree_length_no))+2*clade_sd,lwd=2,lty=2)
abline(v=(mean(temp_tree_length_no))-3*clade_sd,lwd=1,lty=2)
abline(v=(mean(temp_tree_length_no))+3*clade_sd,lwd=1,lty=2)

abline(v=length(unique(node_label_vec_NJ_numbers)),lwd=3,col="blue")
```

According to this test, we can reject the null hypothesis. The number of tips shared between the NJ and host tree is significantly larger than the number of tips shared between randomly-generated trees and the host trees at a threshold of p<0.01.

## Discussion and Future Directions

The tests performed in this exercise measure the similarity between host and two endosymbiont trees, with the NJ tree being somewhat more similar to the host tree than the UPGMA tree. As of now (5/18/2022) I do not yet have a host tree with the tip labels of the endosymbiont tree. Though, I expect the results using an actual host tree to resemble those that were generated in this exercise. Bacteria, Rickettsia included, are known to demonstrate horizontal gene transfer. Not all taxa in the endosymbiont trees have the same amount of data in these trees either, so these endosymbiont trees as they are unlikely to be similar to the host tree anyway. The relatively large number of tips within shared clades between host and endosymbiont trees is strange -- if I had to guess it has something to do with the large loop used to generate the permutated phylogenies.

Therefore, I plan to explore different methods of tree construction to build better endosymbiont trees than the NJ and UPGMA trees I am currently using for this exercise. I've been told analyses in the R package parafitt would be useful anlyze host-parasite phylogenies, and I'm sure there are even more analyses to explore as well. 